src/serial/MidiSessionALSA.cc

```diff
@@ -166,13 +174,253 @@ void MidiOutALSA::serialize(Archive& /*ar*/, unsigned /*version*/)
+	int srcClient;
+	int srcPort;
+	std::string name;
+	std::string desc;
+	bool connected = false;
+	bool stop = false;
+	cb_queue<uint8_t> queue;
+	std::mutex mutex; // to protect queue
+};
+
+MidiInALSA::MidiInALSA(
+		EventDistributor& eventDistributor_,
+		Scheduler& scheduler_,
+		snd_seq_t& seq_,
+		snd_seq_client_info_t& cinfo, snd_seq_port_info_t& pinfo)
+	: eventDistributor(eventDistributor_)
+	, scheduler(scheduler_)
+	, seq(seq_)
+	, srcClient(snd_seq_port_info_get_client(&pinfo))
+	, srcPort(snd_seq_port_info_get_port(&pinfo))
+	, name(snd_seq_client_info_get_name(&cinfo))
+	, desc(snd_seq_port_info_get_name(&pinfo))
+{
+	eventDistributor.registerEventListener(EventType::MIDI_IN_ALSA, *this);
+	if ((snd_seq_port_info_get_capability(&pinfo) & (SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE))) {
+		name.append(" (input)");
+	}
+}
+
+MidiInALSA::~MidiInALSA()
+{
+	if (connected) {
+		disconnect();
+	}
+	eventDistributor.unregisterEventListener(EventType::MIDI_IN_ALSA, *this);
+}
+
+void MidiInALSA::plugHelper(Connector& connector_, EmuTime::param /*time*/)
+{
+	auto& midiConnector = checked_cast<MidiInConnector&>(connector_);
+	midiConnector.setDataBits(SerialDataInterface::DATA_8); // 8 data bits
+	midiConnector.setStopBits(SerialDataInterface::STOP_1); // 1 stop bit
+	midiConnector.setParityBit(false, SerialDataInterface::EVEN); // no parity
+
+	setConnector(&midiConnector); // base class will do this in a moment,
+	                              // but thread already needs it
+	connect();
+}
+
+void MidiInALSA::unplugHelper(EmuTime::param /*time*/)
+{
+	if (connected) {
+		disconnect();
+	}
+}
+
+void MidiInALSA::connect()
+{
+	destinationPort = snd_seq_create_simple_port(
+		&seq, "MIDI in pluggable",
+		SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE, SND_SEQ_PORT_TYPE_MIDI_GENERIC);
+	if (destinationPort < 0) {
+		throw PlugException(
+			"Failed to create ALSA port: ", snd_strerror(destinationPort));
+	}
+
+	int err = snd_seq_connect_from(&seq, destinationPort, srcClient, srcPort);
+	if (err) {
+		snd_seq_delete_simple_port(&seq, destinationPort);
+		throw PlugException(
+			"Failed to connect to ALSA port "
+			"(", srcClient, ':', srcPort, ")"
+			": ", snd_strerror(err));
+	}
+
+	snd_midi_event_new(MidiOutDevice::MAX_MESSAGE_SIZE, &event_parser);
+	snd_midi_event_no_status(event_parser, 1);
+
+	connected = true;
+	stop = false;
+
+	thread = std::thread([this]() { run(); });
+}
+
+void MidiInALSA::disconnect()
+{
+	stop = true;
+	thread.join();
+
+	snd_midi_event_free(event_parser);
+	snd_seq_disconnect_from(&seq, destinationPort, srcClient, srcPort);
+	snd_seq_delete_simple_port(&seq, destinationPort);
+
+	connected = false;
+}
+
+void MidiInALSA::run()
+{
+	assert(isPluggedIn());
+
+	auto npfd = snd_seq_poll_descriptors_count(&seq, POLLIN);
+	std::vector<struct pollfd> pfd(npfd);
+	snd_seq_poll_descriptors(&seq, pfd.data(), npfd, POLLIN);
+
+	while (true) {
+		if (poll(pfd.data(), npfd, 1000) > 0) {
+			do {
```

Comment:
I was confused why the `do { ... } while (0);` loop is needed.
I *think* it's to direct where the 'break' statements in this loop should go to, correct?
Maybe an alternative could be to change those 'breaks' into 'continue' together with moving `if (stop) break;` before `if (poll(...)) {`.

Reply:
It's to break out of the if block without exiting the function entirely, since C/C++ doesn't allow you to use `break` statements outside for and while loops.

Reply:
Yes, hence my suggestion to write it like:
```
while (true) {
    if (stop) break;
    if (poll(...)) {
        if (...) {
            continue;
        }
    }
}
```
But I'll leave it up to you.

Thanks for addressing all my other review comments so quickly.

Hopefully @mthuurne can make time for a functional review in the coming days. If not, then I'll pull it anyway. I trust you tested it well.

Reply:
Done.

Reply:
Done.